File: haskell/src/Euler/Problem/Problem17.hs
module Euler.Problem.Problem17 (answer) where

answer :: Int
answer = sum $ map length _1_1000 where

    _1_9   = [ "one", "two", "three", "four", "five"
             , "six", "seven", "eight", "nine" ]

    _10_19 = [ "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen"
             , "sixteen", "seventeen", "eighteen", "nineteen" ]

    _20_99 = do a <- [ "twenty", "thirty", "forty", "fifty"
                     , "sixty", "seventy", "eighty", "ninety" ]
                b <- "" : _1_9
                return (a ++ b)

    _1_99 = _1_9 ++ _10_19 ++ _20_99

    _100_999 = do a <- map (++ "hundred") _1_9
                  b <- "" : map ("and" ++) _1_99
                  return (a ++ b)

    _1_1000 = _1_99 ++ _100_999 ++ [ "onethousand" ]

File: scala/src/main/scala/euler/problem/Problem17.scala
package euler.problem

object Problem17 {

  lazy val answer: Int = strings.map(_.length).sum

  lazy val strings: Seq[String] = {

    val `1-9` = Seq("one", "two", "three", "four", "five",
                    "six", "seven", "eight", "nine")

    val `10-19` = Seq("ten", "eleven", "twelve", "thirteen",
                      "fourteen", "fifteen", "sixteen", "seventeen",
                      "eighteen", "nineteen")

    val `20-99` =
      for {
        a <- Seq("twenty", "thirty", "forty", "fifty",
                 "sixty", "seventy", "eighty", "ninety")
        b <- "" +: `1-9`
      } yield a + b

    val `1-99` = `1-9` ++ `10-19` ++ `20-99`

    val `100-999` =
      for {
        a <- `1-9`
        b <- "" +: `1-99`.map("and" + _)
      } yield a + "hundred" + b

    val `1-1000` = `1-99` ++ `100-999` :+ "onethousand"

    assert(`1-1000`.size == 1000)

    `1-1000`
  }

}
